### ISSUE TITLE
 systemd-sysupdate: libcurl segfault when updating

### ISSUE THREAD

**Issue by eean** [systemd/systemd#35322]
### systemd version the issue has been seen with

257.rc1.r77601.9466fe014f6-1-arch-g9466fe0^

### Used distribution

KDE Linux 202411170743

### Linux kernel version used

6.11.9-arch1-1

### CPU architectures issue was seen on

None

### Component

other

### Expected behaviour you didn't see

systemd-sysupdate pulls latest efi and root tarballs from https://files.kde.org/kde-linux and updates system

### Unexpected behaviour you saw

systemd-sysupdate SIGABRTs

### Steps to reproduce the problem

Running the update.sh script in https://invent.kde.org/kde-linux/kde-linux runs this command:
  /usr/lib/systemd/systemd-sysupdate --definitions=mkosi.extra/usr/lib/sysupdate.d --verify=no update
which SIGABRTs. 

The curl version is the Arch Linux build 8.11.0-3

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
# coredumpctl gdb 
           PID: 353205 (systemd-pull)
           UID: 0 (root)
           GID: 0 (root)
        Signal: 6 (ABRT)
     Timestamp: Fri 2024-11-22 11:19:15 PST (2min 1s ago)
  Command Line: /usr/lib/systemd/systemd-pull raw --direct --verify no https://files.kde.org/kde-linux/SHA256SUMS -
    Executable: /usr/lib/systemd/systemd-pull
 Control Group: /user.slice/user-0.slice/session-77.scope
          Unit: session-77.scope
         Slice: user-0.slice
       Session: 77
     Owner UID: 0 (root)
       Boot ID: 485cc8ab73234397a29d40f9da963b62
    Machine ID: fcadc390589f4f85b0e4016391602c00
      Hostname: onigiri
       Storage: /var/lib/systemd/coredump/core.systemd-pull.0.485cc8ab73234397a29d40f9da963b62.353205.1732303155000000.zst (present)
  Size on Disk: 643K
       Message: Process 353205 (systemd-pull) of user 0 dumped core.
                
                Stack trace of thread 353205:
                #0  0x00007398f46b33f4 n/a (libc.so.6 + 0x963f4)
                #1  0x00007398f465a120 raise (libc.so.6 + 0x3d120)
                #2  0x00007398f46414c3 abort (libc.so.6 + 0x244c3)
                #3  0x00007398f4a7ac21 log_assert_failed (libsystemd-shared-257.rc1.r77601.9466fe014f6-1.so + 0x7ac21)
                #4  0x00007398f4c15588 log_assert_failed_return (libsystemd-shared-257.rc1.r77601.9466fe014f6-1.so + 0x215588)
                #5  0x00007398f4c97a10 sd_event_add_io (libsystemd-shared-257.rc1.r77601.9466fe014f6-1.so + 0x297a10)
                #6  0x00005cd98c65d4d7 curl_glue_socket_callback (/usr/lib/systemd/systemd-pull + 0xc4d7)
                #7  0x00007398f4ef335b n/a (libcurl.so.4 + 0x5735b)
                #8  0x00007398f4eb391c n/a (libcurl.so.4 + 0x1791c)
                #9  0x00007398f4eba602 n/a (libcurl.so.4 + 0x1e602)
                #10 0x00007398f4ebaf84 n/a (libcurl.so.4 + 0x1ef84)
                #11 0x00007398f4ef45db curl_multi_cleanup (libcurl.so.4 + 0x585db)
                #12 0x00005cd98c65d7e4 curl_glue_unref (/usr/lib/systemd/systemd-pull + 0xc7e4)
                #13 0x00005cd98c6616e5 raw_pull_unref (/usr/lib/systemd/systemd-pull + 0x106e5)
                #14 0x00005cd98c664556 raw_pull_unref (/usr/lib/systemd/systemd-pull + 0x13556)
                #15 0x00005cd98c65603a pull_main (/usr/lib/systemd/systemd-pull + 0x503a)
                #16 0x00007398f4642e08 n/a (libc.so.6 + 0x25e08)
                #17 0x00007398f4642ecc __libc_start_main (libc.so.6 + 0x25ecc)
                #18 0x00005cd98c6565a5 _start (/usr/lib/systemd/systemd-pull + 0x55a5)
                ELF object binary architecture: AMD x86-64

GNU gdb (GDB) 15.2
Copyright (C) 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/lib/systemd/systemd-pull...
[New LWP 353205]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
Core was generated by `/usr/lib/systemd/systemd-pull raw --direct --verify no https://files.kde.org/kd'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007398f46b33f4 in ?? () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007398f46b33f4 in ?? () from /usr/lib/libc.so.6
#1  0x00007398f465a120 in raise () from /usr/lib/libc.so.6
#2  0x00007398f46414c3 in abort () from /usr/lib/libc.so.6
#3  0x00007398f4a7ac21 in log_assert_failed (text=<optimized out>, file=<optimized out>, line=<optimized out>, 
    func=func@entry=0x7398f4dc1ad0 <__func__.141> "sd_event_add_io") at ../sysupdated-systemd-git-stable/src/basic/log.c:996
#4  0x00007398f4c15588 in log_assert_failed_return (text=0x7398f4d9b2aa "e->state != SD_EVENT_FINISHED", 
    file=0x7398f4d74c01 "src/libsystemd/sd-event/sd-event.c", line=1260, func=0x7398f4dc1ad0 <__func__.141> "sd_event_add_io")
    at ../sysupdated-systemd-git-stable/src/basic/log.c:1015
#5  0x00007398f4c97a10 in sd_event_add_io (e=<optimized out>, ret=ret@entry=0x7ffccb0fbd60, fd=fd@entry=9, events=1, 
    callback=callback@entry=0x5cd98c65d110 <curl_glue_on_io>, userdata=userdata@entry=0x5cd9949e53f0)
    at ../sysupdated-systemd-git-stable/src/libsystemd/sd-event/sd-event.c:1260
#6  0x00005cd98c65d4d7 in curl_glue_socket_callback (curl=<optimized out>, s=9, action=1, userdata=0x5cd9949e53f0, 
    socketp=<optimized out>) at ../sysupdated-systemd-git-stable/src/import/curl-util.c:98
#7  0x00007398f4ef335b in ?? () from /usr/lib/libcurl.so.4
#8  0x00007398f4eb391c in ?? () from /usr/lib/libcurl.so.4
#9  0x00007398f4eba602 in ?? () from /usr/lib/libcurl.so.4
#10 0x00007398f4ebaf84 in ?? () from /usr/lib/libcurl.so.4
#11 0x00007398f4ef45db in curl_multi_cleanup () from /usr/lib/libcurl.so.4
#12 0x00005cd98c65d7e4 in curl_glue_unref (g=0x5cd9949e53f0) at ../sysupdated-systemd-git-stable/src/import/curl-util.c:185
#13 0x00005cd98c6616e5 in raw_pull_unref (i=0x5cd9949e5ed0) at ../sysupdated-systemd-git-stable/src/import/pull-raw.c:95
#14 0x00005cd98c664556 in raw_pull_unref (i=0x5cd9949e5ed0) at ../sysupdated-systemd-git-stable/src/import/pull-raw.c:84
#15 raw_pull_unrefp (p=<optimized out>) at ../sysupdated-systemd-git-stable/src/import/pull-raw.h:17
#16 pull_raw (argc=<optimized out>, argv=<optimized out>, userdata=<optimized out>)
    at ../sysupdated-systemd-git-stable/src/import/pull.c:187
#17 0x00005cd98c65603a in pull_main (argc=7, argv=0x7ffccb0fc448) at ../sysupdated-systemd-git-stable/src/import/pull.c:576
#18 run (argc=7, argv=0x7ffccb0fc448) at ../sysupdated-systemd-git-stable/src/import/pull.c:593
#19 main (argc=7, argv=0x7ffccb0fc448) at ../sysupdated-systemd-git-stable/src/import/pull.c:596
(gdb) quit
```

**Comment by bluca** [comment]
Are you able to reproduce with the debug symbols for libcurl installed so that the backtrace is complete?

**Comment by yuwata** [comment]
fix is waiting in #35327.

**Comment by eean** [comment]
Wow, 10 hour turnaround, thank you @yuwata !
